README.md
# bluetooth_thermal_printer

[![Pub Version](https://img.shields.io/pub/v/bluetooth_thermal_printer)](https://pub.dev/packages/bluetooth_thermal_printer)

The library allows printing receipts using a Bluetooth printer(Android Only).  It supports both 58mm and 80mm Bluetooth printers. 

This package does not use Location Permission. By that, it follows google's policy of Android 10.

## Add-ons
1. [esc_pos_utils](https://pub.dev/packages/esc_pos_utils) package to print receipts 
2. [Image](https://pub.dev/packages/image) to print images


### Simple Ticket with Styles:
```dart
List<int> testTicket() {
  final List<int> bytes = [];
  // Using default profile
  final profile = await CapabilityProfile.load();
  final generator = Generator(PaperSize.mm80, profile);
  List<int> bytes = [];

  bytes += generator.text(
      'Regular: aA bB cC dD eE fF gG hH iI jJ kK lL mM nN oO pP qQ rR sS tT uU vV wW xX yY zZ');
  bytes += generator.text('Special 1: àÀ èÈ éÉ ûÛ üÜ çÇ ôÔ',
      styles: PosStyles(codeTable: PosCodeTable.westEur));
  bytes += generator.text('Special 2: blåbærgrød',
      styles: PosStyles(codeTable: PosCodeTable.westEur));

  bytes += generator.text('Bold text', styles: PosStyles(bold: true));
  bytes += generator.text('Reverse text', styles: PosStyles(reverse: true));
  bytes += generator.text('Underlined text',
      styles: PosStyles(underline: true), linesAfter: 1);
  bytes += generator.text('Align left', styles: PosStyles(align: PosAlign.left));
  bytes += generator.text('Align center', styles: PosStyles(align: PosAlign.center));
  bytes += generator.text('Align right',
      styles: PosStyles(align: PosAlign.right), linesAfter: 1);

  bytes += generator.text('Text size 200%',
      styles: PosStyles(
        height: PosTextSize.size2,
        width: PosTextSize.size2,
      ));

  bytes += generator.feed(2);
  bytes += generator.cut();
  return bytes;
}
```

You can find more examples here: [esc_pos_utils](https://github.com/andrey-ushakov/esc_pos_utils).

## How to Help
* Test your printer and add it in the table: [Tested Printers](printers.md)
* Test and report bugs
* Share your ideas about what could be improved (code optimization, new features...)
* PRs are welcomed!


## Tested Printers
Here are some [printers tested with this library](printers.md). Please add the models you have tested to maintain and improve this library and help others to choose the right printer.

For a complete example, check the [example project](example/).

## Test Print
<img src="https://raw.githubusercontent.com/Rajath4/BluetoothThermalPrinter/master/demo1.jpg" alt="test receipt" width="400"/>

<img src="https://raw.githubusercontent.com/Rajath4/BluetoothThermalPrinter/master/demo2.jpg" alt="test receipt" width="400"/>

CHANGELOG.md
## 0.0.6

* Updated description

## 0.0.5

* Added Generator in example and flutter 2.0 support

## 0.0.4

* Updated document


## 0.0.3

* Updated document


## 0.0.2

* Updated document


## 0.0.1

* Initial release.







example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:bluetooth_thermal_printer_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/bluetooth_thermal_printer_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:bluetooth_thermal_printer/bluetooth_thermal_printer.dart';

void main() {
  const MethodChannel channel = MethodChannel('bluetooth_thermal_printer');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await BluetoothThermalPrinter.platformVersion, '42');
  });
}


example/lib/main.dart
import 'dart:async';

import 'package:bluetooth_thermal_printer/bluetooth_thermal_printer.dart';
import 'package:esc_pos_utils/esc_pos_utils.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  bool connected = false;
  List availableBluetoothDevices = [];

  Future<void> getBluetooth() async {
    final List bluetooths = await BluetoothThermalPrinter.getBluetooths;
    print("Print $bluetooths");
    setState(() {
      availableBluetoothDevices = bluetooths;
    });
  }

  Future<void> setConnect(String mac) async {
    final String result = await BluetoothThermalPrinter.connect(mac);
    print("state conneected $result");
    if (result == "true") {
      setState(() {
        connected = true;
      });
    }
  }

  Future<void> printTicket() async {
    String isConnected = await BluetoothThermalPrinter.connectionStatus;
    if (isConnected == "true") {
      List<int> bytes = await getTicket();
      final result = await BluetoothThermalPrinter.writeBytes(bytes);
      print("Print $result");
    } else {
      //Hadnle Not Connected Senario
    }
  }

  Future<void> printGraphics() async {
    String isConnected = await BluetoothThermalPrinter.connectionStatus;
    if (isConnected == "true") {
      List<int> bytes = await getGraphicsTicket();
      final result = await BluetoothThermalPrinter.writeBytes(bytes);
      print("Print $result");
    } else {
      //Hadnle Not Connected Senario
    }
  }

  Future<List<int>> getGraphicsTicket() async {
    List<int> bytes = [];

    CapabilityProfile profile = await CapabilityProfile.load();
    final generator = Generator(PaperSize.mm80, profile);

    // Print QR Code using native function
    bytes += generator.qrcode('example.com');

    bytes += generator.hr();

    // Print Barcode using native function
    final List<int> barData = [1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 4];
    bytes += generator.barcode(Barcode.upcA(barData));

    bytes += generator.cut();

    return bytes;
  }

  Future<List<int>> getTicket() async {
    List<int> bytes = [];
    CapabilityProfile profile = await CapabilityProfile.load();
    final generator = Generator(PaperSize.mm80, profile);

    bytes += generator.text("Demo Shop",
        styles: PosStyles(
          align: PosAlign.center,
          height: PosTextSize.size2,
          width: PosTextSize.size2,
        ),
        linesAfter: 1);

    bytes += generator.text(
        "18th Main Road, 2nd Phase, J. P. Nagar, Bengaluru, Karnataka 560078",
        styles: PosStyles(align: PosAlign.center));
    bytes += generator.text('Tel: +919591708470',
        styles: PosStyles(align: PosAlign.center));

    bytes += generator.hr();
    bytes += generator.row([
      PosColumn(
          text: 'No',
          width: 1,
          styles: PosStyles(align: PosAlign.left, bold: true)),
      PosColumn(
          text: 'Item',
          width: 5,
          styles: PosStyles(align: PosAlign.left, bold: true)),
      PosColumn(
          text: 'Price',
          width: 2,
          styles: PosStyles(align: PosAlign.center, bold: true)),
      PosColumn(
          text: 'Qty',
          width: 2,
          styles: PosStyles(align: PosAlign.center, bold: true)),
      PosColumn(
          text: 'Total',
          width: 2,
          styles: PosStyles(align: PosAlign.right, bold: true)),
    ]);

    bytes += generator.row([
      PosColumn(text: "1", width: 1),
      PosColumn(
          text: "Tea",
          width: 5,
          styles: PosStyles(
            align: PosAlign.left,
          )),
      PosColumn(
          text: "10",
          width: 2,
          styles: PosStyles(
            align: PosAlign.center,
          )),
      PosColumn(text: "1", width: 2, styles: PosStyles(align: PosAlign.center)),
      PosColumn(text: "10", width: 2, styles: PosStyles(align: PosAlign.right)),
    ]);

    bytes += generator.row([
      PosColumn(text: "2", width: 1),
      PosColumn(
          text: "Sada Dosa",
          width: 5,
          styles: PosStyles(
            align: PosAlign.left,
          )),
      PosColumn(
          text: "30",
          width: 2,
          styles: PosStyles(
            align: PosAlign.center,
          )),
      PosColumn(text: "1", width: 2, styles: PosStyles(align: PosAlign.center)),
      PosColumn(text: "30", width: 2, styles: PosStyles(align: PosAlign.right)),
    ]);

    bytes += generator.row([
      PosColumn(text: "3", width: 1),
      PosColumn(
          text: "Masala Dosa",
          width: 5,
          styles: PosStyles(
            align: PosAlign.left,
          )),
      PosColumn(
          text: "50",
          width: 2,
          styles: PosStyles(
            align: PosAlign.center,
          )),
      PosColumn(text: "1", width: 2, styles: PosStyles(align: PosAlign.center)),
      PosColumn(text: "50", width: 2, styles: PosStyles(align: PosAlign.right)),
    ]);

    bytes += generator.row([
      PosColumn(text: "4", width: 1),
      PosColumn(
          text: "Rova Dosa",
          width: 5,
          styles: PosStyles(
            align: PosAlign.left,
          )),
      PosColumn(
          text: "70",
          width: 2,
          styles: PosStyles(
            align: PosAlign.center,
          )),
      PosColumn(text: "1", width: 2, styles: PosStyles(align: PosAlign.center)),
      PosColumn(text: "70", width: 2, styles: PosStyles(align: PosAlign.right)),
    ]);

    bytes += generator.hr();

    bytes += generator.row([
      PosColumn(
          text: 'TOTAL',
          width: 6,
          styles: PosStyles(
            align: PosAlign.left,
            height: PosTextSize.size4,
            width: PosTextSize.size4,
          )),
      PosColumn(
          text: "160",
          width: 6,
          styles: PosStyles(
            align: PosAlign.right,
            height: PosTextSize.size4,
            width: PosTextSize.size4,
          )),
    ]);

    bytes += generator.hr(ch: '=', linesAfter: 1);

    // ticket.feed(2);
    bytes += generator.text('Thank you!',
        styles: PosStyles(align: PosAlign.center, bold: true));

    bytes += generator.text("26-11-2020 15:22:45",
        styles: PosStyles(align: PosAlign.center), linesAfter: 1);

    bytes += generator.text(
        'Note: Goods once sold will not be taken back or exchanged.',
        styles: PosStyles(align: PosAlign.center, bold: false));
    bytes += generator.cut();
    return bytes;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Bluetooth Thermal Printer Demo'),
        ),
        body: Container(
          padding: EdgeInsets.all(20),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text("Search Paired Bluetooth"),
              TextButton(
                onPressed: () {
                  this.getBluetooth();
                },
                child: Text("Search"),
              ),
              Container(
                height: 200,
                child: ListView.builder(
                  itemCount: availableBluetoothDevices.length > 0
                      ? availableBluetoothDevices.length
                      : 0,
                  itemBuilder: (context, index) {
                    return ListTile(
                      onTap: () {
                        String select = availableBluetoothDevices[index];
                        List list = select.split("#");
                        // String name = list[0];
                        String mac = list[1];
                        this.setConnect(mac);
                      },
                      title: Text('${availableBluetoothDevices[index]}'),
                      subtitle: Text("Click to connect"),
                    );
                  },
                ),
              ),
              SizedBox(
                height: 30,
              ),
              TextButton(
                onPressed: connected ? this.printGraphics : null,
                child: Text("Print"),
              ),
              TextButton(
                onPressed: connected ? this.printTicket : null,
                child: Text("Print Ticket"),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/bluetooth_thermal_printer.dart
import 'dart:async';

import 'package:flutter/services.dart';

class BluetoothThermalPrinter {
  static const MethodChannel _channel =
      const MethodChannel('bluetooth_thermal_printer');

  /// Get android platform version.
  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  /// Get list of all paired bluetooth devices
  static Future<List> get getBluetooths async {
    List items = new List();
    try {
      final List result = await _channel.invokeMethod('bluetothLinked');
      items = result;
    } on PlatformException catch (e) {
      print("Bluetooth paired failure: '${e.message}'.");
    }

    return items;
  }

  /// Get bluetooth device connection status
  static Future<String> get connectionStatus async {
    try {
      final String result = await _channel.invokeMethod('connectionStatus');
      return result;
    } on PlatformException catch (e) {
      print("Failed to write bytes: '${e.message}'.");
      return "false";
    }
  }

  /// Connect to device using [mac].
  static Future<String> connect(String mac) async {
    String result = "false";
    try {
      result = await _channel.invokeMethod('connectPrinter', mac);
    } on PlatformException catch (e) {
      print("Failed to connect: '${e.message}'.");
    }
    return result;
  }

  ///Printes the [bytes] using bluetooth printer.
  static Future<String> writeBytes(List<int> bytes) async {
    try {
      final String result = await _channel.invokeMethod('writeBytes', bytes);
      return result;
    } on PlatformException catch (e) {
      print("Failed to write bytes: '${e.message}'.");
      return "false";
    }
  }

  ///Printes the [text] using bluetooth printer.
  static Future<String> writeText(String text) async {
    try {
      final String result = await _channel.invokeMethod('printText', text);
      return result;
    } on PlatformException catch (e) {
      print("Failed to writeText: '${e.message}'.");
      return "false";
    }
  }

  /// Get battery level of the android device.
  static Future<int> get getBatteryLevel async {
    try {
      final int result = await _channel.invokeMethod('getBatteryLevel');
      return result;
    } on PlatformException catch (e) {
      print("Failed to get battery level: '${e.message}'.");
      return -1;
    }
  }
}


android/src/main/kotlin/com/peoplewareinnovations/bluetooth_thermal_printer/BluetoothThermalPrinterPlugin.kt
package com.peoplewareinnovations.bluetooth_thermal_printer

import android.app.Activity
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.content.Context
import android.content.ContextWrapper
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import android.util.Log
import android.widget.Toast
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.OutputStream
import java.util.*
import androidx.annotation.NonNull


private const val TAG = "====> mio: "
private var outputStream: OutputStream? = null
private lateinit var mac: String
//val REQUEST_ENABLE_BT = 2

class BluetoothThermalPrinterPlugin: FlutterPlugin, MethodCallHandler{
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var mContext: Context
  private lateinit var channel : MethodChannel
  private lateinit var state:String

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "bluetooth_thermal_printer")
    channel.setMethodCallHandler(this)
    this.mContext = flutterPluginBinding.applicationContext
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    }else if (call.method == "getBatteryLevel") {
      val batteryLevel = getBatteryLevel()
      if (batteryLevel != -1) {
        result.success(batteryLevel)
      } else {
        result.error("UNAVAILABLE", "Battery level not available.", null)
      }
    }else if (call.method == "BluetoothStatus") {
      var state:String = "false"
      val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
      if (bluetoothAdapter != null && bluetoothAdapter.isEnabled) {
        state = "true"
      }
      result.success(state)
    }else if (call.method == "connectionStatus") {

      if(outputStream != null) {
        try{
          outputStream?.run {
            write(" ".toByteArray())
            result.success("true")
            //Log.d(TAG, "step yes connection")
          }
        }catch (e: Exception){
          result.success("false")
          outputStream = null
          ShowToast("Device was disconnected, reconnect")
          //Log.d(TAG, "state print: ${e.message}")
        }
      }else{
        result.success("false")
        //Log.d(TAG, "no paso es false ")
      }

    } else if (call.method == "connectPrinter") {
      var printerMAC = call.arguments.toString();
      if(printerMAC.length>0){
        mac = printerMAC;
      }else{
        result.success("false")
      }
      GlobalScope.launch(Dispatchers.Main) {
        if(outputStream == null) {
          outputStream = connect()?.also {
            //result.success("true")
            //Toast.makeText(this@MainActivity, "Connected to printer", Toast.LENGTH_SHORT).show()
          }.apply {
            result.success(state)
            //Log.d(TAG, "finished: Connection state:$state")
          }
        }
      }
    }else if (call.method == "writeBytes") {

      var lista: List<Int> = call.arguments as List<Int>
      var bytes: ByteArray = "\n".toByteArray()

      lista.forEach {
        bytes += it.toByte() //Log.d(TAG, "foreach: ${it}")
      }
      if(outputStream != null) {
        try{
          outputStream?.run {
            write(bytes)
            result.success("true")
          }
        }catch (e: Exception){
          result.success("false")
          outputStream = null
          ShowToast("Device was disconnected, reconnect")
          // Log.d(TAG, "state print: ${e.message}")
          /*var ex:String = e.message.toString()
          if(ex=="Broken pipe"){
            Log.d(TAG, "Device was disconnected, reconnect")
            ShowToast("Device was disconnected, reconnect")
          }*/
        }
      }else{
        result.success("false")
      }

    }else if (call.method == "printText") {

      var stringArrived: String = call.arguments.toString()
      //var list = stringArrived.split("*")
      //println("list ${list.toString()}")

      if(outputStream != null) {
        try{
          var size:Int = 0
          var texto:String = ""
          var line = stringArrived.split("//")
          //Log.d(TAG, "list arrived: ${line.size}")
          if(line.size>1) {
            size = line[0].toInt()
            texto = line[1]
            if (size < 1 || size > 5) size = 2
          }else{
            size = 2
            texto = stringArrived
            //Log.d(TAG, "list came 2 text: ${texto} size: $size")
          }

          outputStream?.run {
            write(setBytes.size[0])
            write(setBytes.cancelar_chino)
            write(setBytes.caracteres_escape)
            write(setBytes.size[size])
            write(texto.toByteArray(charset("iso-8859-1")))
            result.success("true")
          }
        }catch (e: Exception){
          result.success("false")
          outputStream = null
          ShowToast("Device was disconnected, reconnect")
        }
      }else{
        result.success("false")
      }

    }else if (call.method == "bluetothLinked") {

      var list:List<String> = getLinkedDevices()

      result.success(list)

    }else {
      result.notImplemented()
    }
  }

  private fun getBatteryLevel(): Int {
    val batteryLevel: Int
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      val batteryManager = mContext?.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
      batteryLevel = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
    } else {
      val intent = ContextWrapper(mContext?.applicationContext).registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
      batteryLevel = intent!!.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) * 100 / intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
    }

    return batteryLevel
  }

  private suspend fun connect(): OutputStream? {
    state = "false"
    return withContext(Dispatchers.IO) {
      var outputStream: OutputStream? = null
      val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
      if (bluetoothAdapter != null && bluetoothAdapter.isEnabled) {
        try {
          val bluetoothAddress = mac//"66:02:BD:06:18:7B" // replace with your device's address
          val bluetoothDevice = bluetoothAdapter.getRemoteDevice(bluetoothAddress)
          val bluetoothSocket = bluetoothDevice?.createRfcommSocketToServiceRecord(
                  UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
          )
          bluetoothAdapter.cancelDiscovery()
          bluetoothSocket?.connect()
          if (bluetoothSocket!!.isConnected) {
            outputStream = bluetoothSocket!!.outputStream
            state = "true"
            //outputStream.write("\n".toByteArray())
          }else{
            state = "false"
            Log.d(TAG, "Disconnected: ")
          }
          //bluetoothSocket?.close()
        } catch (e: Exception){
          state = "false"
          var code:Int = e.hashCode() //1535159 off //
          Log.d(TAG, "connect: ${e.message} code $code")
          outputStream?.close()
        }
      }else{
        state = "false"
        Log.d(TAG, "Adapter problem")
      }
      outputStream
    }
  }

  private fun getLinkedDevices():List<String>{

    val listItems: MutableList<String> = mutableListOf()

    val bluetoothAdapter: BluetoothAdapter? = BluetoothAdapter.getDefaultAdapter()
    if (bluetoothAdapter == null) {
      //lblmsj.setText("This application needs a phone with bluetooth")
    }
    //if blue tooth is not on
    if (bluetoothAdapter?.isEnabled == false) {
      //val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
      //startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT)
      //ShowToast("Bluetooth off")
    }
    //search bluetooth
    //Log.d(TAG, "searching for devices: ")
    val pairedDevices: Set<BluetoothDevice>? = bluetoothAdapter?.bondedDevices
    pairedDevices?.forEach { device ->
      val deviceName = device.name
      val deviceHardwareAddress = device.address
      listItems.add("$deviceName#$deviceHardwareAddress")
      //Log.d(TAG, "device: ${device.name}")
    }

    return listItems;
  }

  private fun ShowToast(message: String){
    Toast.makeText(mContext, message, Toast.LENGTH_SHORT).show()
  }

  class setBytes(){
    companion object {
      //val info = "This is info"
      //fun getMoreInfo():String { return "This is more fun" }

      val enter = "\n".toByteArray()
      val resetear_impresora = byteArrayOf(0x1b, 0x40, 0x0a)
      val cancelar_chino = byteArrayOf(0x1C, 0x2E)
      val caracteres_escape = byteArrayOf(0x1B, 0x74, 0x10)

      val size = arrayOf(
              byteArrayOf(0x1d, 0x21, 0x00), // La fuente no se agranda 0
              byteArrayOf(0x1b, 0x4d, 0x01), // Fuente ASCII comprimida 1
              byteArrayOf(0x1b, 0x4d, 0x00), //Fuente estándar ASCII    2
              byteArrayOf(0x1d, 0x21, 0x11), // Altura doblada 3
              byteArrayOf(0x1d, 0x21, 0x22), // Altura doblada 4
              byteArrayOf(0x1d, 0x21, 0x33) // Altura doblada 5
      )


      //deprecated codes
      const val HT: Byte = 9
      const val LF: Byte = 10
      const val CR: Byte = 13
      const val ESC: Byte = 27
      const val DLE: Byte = 16
      const val GS: Byte = 29
      const val FS: Byte = 28
      const val STX: Byte = 2
      const val US: Byte = 31
      const val CAN: Byte = 24
      const val CLR: Byte = 12
      const val EOT: Byte = 4
      val INIT = byteArrayOf(27, 64)
      var FEED_LINE = byteArrayOf(10)
      var SELECT_FONT_A = byteArrayOf(20, 33, 0)
      var SET_BAR_CODE_HEIGHT = byteArrayOf(29, 104, 100)
      var PRINT_BAR_CODE_1 = byteArrayOf(29, 107, 2)
      var SEND_NULL_BYTE = byteArrayOf(0)
      var SELECT_PRINT_SHEET = byteArrayOf(27, 99, 48, 2)
      var FEED_PAPER_AND_CUT = byteArrayOf(29, 86, 66, 0)
      var SELECT_CYRILLIC_CHARACTER_CODE_TABLE = byteArrayOf(27, 116, 17)
      var SELECT_BIT_IMAGE_MODE = byteArrayOf(27, 42, 33, -128, 0)
      var SET_LINE_SPACING_24 = byteArrayOf(27, 51, 24)
      var SET_LINE_SPACING_30 = byteArrayOf(27, 51, 30)
      var TRANSMIT_DLE_PRINTER_STATUS = byteArrayOf(16, 4, 1)
      var TRANSMIT_DLE_OFFLINE_PRINTER_STATUS = byteArrayOf(16, 4, 2)
      var TRANSMIT_DLE_ERROR_STATUS = byteArrayOf(16, 4, 3)
      var TRANSMIT_DLE_ROLL_PAPER_SENSOR_STATUS = byteArrayOf(16, 4, 4)
      val ESC_FONT_COLOR_DEFAULT = byteArrayOf(27, 114, 0)
      val FS_FONT_ALIGN = byteArrayOf(28, 33, 1, 27, 33, 1)
      val ESC_ALIGN_LEFT = byteArrayOf(27, 97, 0)
      val ESC_ALIGN_RIGHT = byteArrayOf(27, 97, 2)
      val ESC_ALIGN_CENTER = byteArrayOf(27, 97, 1)
      val ESC_CANCEL_BOLD = byteArrayOf(27, 69, 0)
      val ESC_HORIZONTAL_CENTERS = byteArrayOf(27, 68, 20, 28, 0)
      val ESC_CANCLE_HORIZONTAL_CENTERS = byteArrayOf(27, 68, 0)
      val ESC_ENTER = byteArrayOf(27, 74, 64)
      val PRINTE_TEST = byteArrayOf(29, 40, 65)
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

